iconservice/icon_service_engine.py

```diff
@@ -591,13 +537,15 @@ def _update_productivity(context: 'IconScoreContext',
         if prev_block_validators:
             validates.update(prev_block_validators)
 
-        main_preps: list = context.engine.prep.term.main_preps
-        for main_prep in main_preps:
-            is_validate: bool = main_prep.address in validates
-            prep: 'PRep' = context.get_prep(main_prep.address, mutable=True)
-            if prep:
-                prep.update_productivity(is_validate)
-                context.put_dirty_prep(prep)
+        main_preps: List['Address'] = context.storage.meta.get_last_main_preps(context)
+
+        for address in main_preps:
+            is_validate: bool = address in validates
+            dirty_prep: 'PRep' = context.get_prep(address, mutable=True)
```

Comment:
I guess None type data could be returned from the get_prep. (should change the type hint to Optional)